crates/validation/src/derivation.rs

```diff
@@ -537,3 +543,30 @@ fn walk_derive_transform(
 
     Some((request, shuffle_types))
 }
+
+fn extract_validated(
+    response: anyhow::Result<derive::Response>,
+) -> Result<(derive::response::Validated, Vec<flow::NetworkPort>), Error> {
+    let response = match response {
+        Ok(response) => response,
+        Err(err) => return Err(Error::Connector { detail: err }),
+    };
+
+    let internal = match response.get_internal() {
+        Ok(internal) => internal,
+        Err(err) => {
+            return Err(Error::Connector {
+                detail: anyhow::anyhow!("parsing internal: {err}"),
+            });
+        }
+    };
+
+    let Some(validated) = response.validated else {
+        return Err(Error::Connector {
+            detail: anyhow::anyhow!("expected Validated but got {}", serde_json::to_string(&response).unwrap()),
+        });
+    };
+    let network_ports = internal.container.unwrap_or_default().network_ports;
```

Comment:
A change here, is that we identify the container IP and then send that back along the `Opened` path to the rest of the runtime, for it to directly call into the container's networking stack. I did this because it a) works on linux docker and production and b) sets things up for Firecracker, which _only_ has an IP and doesn't have any built-in facilities for port forwarding.

But. I'm just realizing this won't work on Docker Desktop for Mac, because these container IPs are only accessible _inside_ the docker VM. Which breaks `flowctl` on Mac. Still thinking through how to handle this....

Reply:
Okay, I think I've worked out a good pathway here:

When running on non-linux, we pass --publish-all as well as --publish 0.0.0.0:0:{CONNECTOR_INIT_PORT} to docker. That causes it to pick some random host ports and publish all image ports through them, including the connector-init port. Next we inspect the container, and, in addition to grabbing the container IP, we also grab any host port mappings a populate a mapping of container-port => host-port-ip:host-port addresses which are also attached to the `runtime::Container` description. Then, if you want to reach a container port, you first look for a mapped host-port and preferentially use that (non-production setups like Mac), otherwise you use the container IP directly.

Next problem: we need an amd64 copy of `flow-connector-init`. The solution I worked up is to use a pinned version of the Flow docker image, from which we pluck out the binary and copy it into a temp location. This is a _bit_ of a kludge -- mostly because it requires pulling down the whole flow image and updating the pin is currently awkward -- but I do think the practice of shipping a vendor'd docker image for `flowctl` to use has legs to it, much like the supabase CLI does.

Tested on my macbook air M2 and working well.